src/app/component/AddressAutocomplete.jsx

```diff
@@ -0,0 +1,295 @@
+    };
+  }
+  return {
+    options: [],
+    zoom: 0,
+    address: "",
+    lat: 0,
+    lng: 0,
+    input: "",
+  };
+};
+function locationReducer(state, action) {
+  switch (action.type) {
+    // input new address, no location yet
+    case "clearOptions": {
+      return { ...state, options: [] };
+    }
+    case "setOptions": {
+      return { ...state, options: action.options };
+    }
+    case "setInput":
+      return { ...state, input: action.input };
+    // once we know the address is valid, we should have new location here
+    case "setLocation":
+      let zoom = state.zoom;
+      if (zoom === 0) {
+        zoom = 16;
+      }
+      return {
+        ...state,
+        zoom: zoom,
+        lat: action.lat,
+        lng: action.lng,
+        address: action.address,
+        input: action.address,
+      };
+    case "reset":
+      return initLocationState();
+    default:
+      throw new Error();
+  }
+}
+
+const autocompleteService = { current: null };
+const geocoderService = { current: null };
+
+/**
+ *
+ * @param {address: "string", lat: number, lng: number} defaultLocation,
+ * refer to Location in scheme.jsx, this location will just show up in the map
+ *  if it is defined, even if the address and the marker are mismatched.
+ *  Once a new address are input though the marker and the address should be correct once again
+ *
+ */
+export default function AddressAutocomplete({
+  defaultLocation,
+  disabled,
+  error,
+  label,
+  onChangeLocation,
+  required,
+  showMap,
+}) {
+  const classes = useStyles();
+  const [location, dispatchLocation] = useReducer(
+    locationReducer,
+    initLocationState(defaultLocation)
+  );
+
+  const loaded = React.useRef(false);
+  loadGoogleMapScript(loaded);
+
+  const handleInputChange = (_, value, reason) => {
+    if (reason === "input") {
+      dispatchLocation({
+        type: "setInput",
+        input: value,
+      });
+    } else {
+      dispatchLocation({ type: "reset" });
+    }
+  };
+  const handleAutoChange = (request) => {
+    let address = "";
+    let lat = 0;
+    let lng = 0;
+
+    if (geocoderService.current && request) {
+      fetchLatLng({ ...request }, (place) => {
+        if (place[0]) {
+          address = place[0].formatted_address;
+          lat = place[0].geometry.location.lat();
+          lng = place[0].geometry.location.lng();
+          dispatchLocation({
+            type: "setLocation",
+            address,
+            lat,
+            lng,
+          });
+          onChangeLocation({
+            address,
+            lat,
+            lng,
+          });
+        }
+      });
+    } else {
+      dispatchLocation({ type: "reset" });
+      onChangeLocation({
+        address,
+        lat,
+        lng,
+      });
+    }
+  };
+
+  const fetchLatLng = useMemo(
+    () =>
+      throttle((request, callback) => {
+        geocoderService.current.geocode(request, callback);
+      }, 200),
+    []
+  );
+  const fetch = useMemo(
```

Comment:
Might want to use a different name here so it's not confused with global `fetch`

Reply:
done